### PR TITLE
Unify SVGs for HTML safe titles reporting encoding errors in linting

### PR DESCRIPTION
**Pull request by mondeja**
> You can see this in action in the [CI linting run of the PR first commit](https://github.com/mondeja/simple-icons/runs/3608971822).

I've found adding tests in #6505 that our SVG files contents are not properly encoded to be HTML safe. Some files like `aeromexico.svg` have unicode non ASCII characters in their titles (in this case `é`). These characters must be encoded using a standard.

For this implementation I've chosen HTML entities decimal number representation (`&#[0-9]+;`) for this unification. I could have chosen:

- Hexadecimal representation (`&x[A-Za-z0-9]+;`), but is more expensive to parse it.
- Named entities (`&[A-Za-z0-9];`), but is more expensive to parse it, would create larger files, and I'm not sure if all unicode characters have their correspondent named entity relation, 

Also, I've to include a mapping file for named entities fixes because I've not found a compatible solution for us already in npm registries after trying with  [`html-entities`](https://www.npmjs.com/package/html-entities) (not exporting data) and [`all-named-html-entities`](https://www.npmjs.com/package/all-named-html-entities) (ES6 only).

### Additional notes

- Let's Encrypt uses the character `’` (`&#8217;`) instead of `'` (`&#39;`) in `data/simple-icons.json`. This may be changed, but would be a breaking change.

### CODE REVIEW

**Comment by ericcornelissen** [comment]
> Let's Encrypt uses the character ’ (&#8217;) instead of ' (&#39;) in data/simple-icons.json. This may be changed, but would be a breaking change.

I'm 99% sure that we should keep Let's Encrypt as it is - at least when it was added that character was the preferred "spelling" of the name.

**Review by ericcornelissen — COMMENTED** [review]
This one's quite a doozy... First off, I believe it's more common for websites nowadays to use UTF8 encoding than ASCII encoding, so from that perspective I personally think we should base this on UTF8 encoding rather than ASCII - though it might still be useful to include a note in our documentation that we make this assumption. I don't know if that changes much about the changes in this PR though...

Second, do you have a source for why _all_ characters have to be escaped? I'm by no means an expert on character encoding, but I couldn't find anything conclusive on whether or not _all_ accented characters have to be escaped. Now, if some **have** to be escaped and others not, we might choose to go the easy route and just escape them all, but to me that's not a given as it decreased the readability of the source code.

> Also, I've to include a mapping file for named entities fixes because I've not found a compatible solution for us already in npm registries after trying with html-entities (not exporting data) and `all-named-html-entities` (ES6 only).

I think it could be worth publishing a non-ES6 version of `all-named-html-entities` (could be just a package that transpiles `all-named-html-entities` and publishes the result :upside_down_face:) ourselves or even for us to upgrade our scripts to ES6 to be able to support this. Though the latter would be out of scope for this PR.

File: icons/atandt.svg
Comment (mondeja) [reply icons/atandt.svg:1]: > I would personally prefer to make an exception for `&` and encode it as `&amp;`  for "readability" (same for other well-supported named entities)

This seems contradictory with the source that you've shared in another comment, quoting textually:

> This is what I found: https://webmasters.stackexchange.com/questions/12335/should-i-escape-the-apostrophe-character-with-its-html-entity-39

> `&apos;` is not a valid HTML character entity reference. It is an XML character entity reference. While Firefox and Chrome, at least, will render the above as an apostrophe in an HTML document, Internet Explorer will not. And it is following the standard when it refuses to do so.

The same happens for `&amp;`, is not a valid HTML entity is an XML one.

Comment (ericcornelissen) [reply icons/atandt.svg:1]: > The same happens for &amp;, is not a valid HTML entity is an XML one.

If that's the case, I agree, but I'm not seeing that being said in that StackExchange thread 🤔 In fact, [this answer](https://webmasters.stackexchange.com/a/12343) seems to explicitly state only `&apos;` is a problem.

**Comment by mondeja** [comment]
> First off, I believe it's more common for websites nowadays to use UTF8 encoding than ASCII encoding, so from that perspective I personally think we should base this on UTF8 encoding rather than ASCII

I think so, but are we targeting websites exclusively? More formally, are we targeting UTF-8 encoded based software? What happens with some people stucked in some other encodings, or in other contexts rather than websites? At first I've implemented it for UTF-8, but I thought it was more robust to stick to ASCII for compatibility concerns.

> I don't know if that changes much about the changes in this PR though...

Just a number in the `127 >=<` loops, there is no problem at all.

> Second, do you have a source for why all characters have to be escaped? I'm by no means an expert on character encoding, but I couldn't find anything conclusive on whether or not all accented characters have to be escaped...

There is no source because it depends how many encodings do you want "to support". There is no such thing of "what characters should be escaped or should not", if you are assuming UTF-8 you can even remove the function `titleToHtmlFriendly`, you don't need nothing of this, except for strange UTF-16 characters. It is a matter of "what encoding are you using"? If your website is UTF-8 encoded you can use the 100% of accented characters in included SVGs without escaping them.

The real question is: how many bytes do you want to ~~"escape"~~ encode (apart from `'&"<>`, of course)? 

- UTF-8:
  - 1 byte: Standard ASCII
  - 2 bytes: Arabic, Hebrew, most European scripts (most notably excluding Georgian)
  - 3 bytes: BMP
  - 4 bytes: All Unicode characters
- UTF-16:
  - 2 bytes: BMP
  - 4 bytes: All Unicode characters

Imagine if I'm a software developer that only wants to support ASCII (one byte UTF-8). This is no a crazy assumption because if you want the fastest SVG path parser in the world, you must forget about other encodings (in fact, you only need a subset of ASCII), assuming that all passed characters are valid ASCII. If you pass a unencoded string to his program for whatever reason, bad things will start to happen. Of course, this is an edge case, but I've had to deal with such decisions in the past.

Another basic example, you want a fully optimized ASCII encoded website:

This code works:
```html
<html>
  <head>
    <meta charset="ascii">
  </head>
  <body>
    <p>Hell&#243;</p>
  </body>
</html>
```
but not this
```html
<html>
  <head>
    <meta charset="ascii">
  </head>
  <body>
    <p>Helló</p>
  </body>
</html>
```

That being said, I am more inclined to the preponderance of universal compatibility over "readability of code" when the code is a basic SVG with two readable tags inside and all created from the same basic template. We want a universal solution or a conventional for western zones? We want to support websites with ASCII encoding or other possible non UTF-8 softwares?

And, if we choose to assume 4 bytes UTF-8, keep in mind that we only need to ~~"escape"~~ encode XML reference characters (`'&<>"`).

> I think it could be worth publishing a non-ES6 version of `all-named-html-entities` (could be just a package that transpiles `all-named-html-entities` and publishes the result :upside_down_face:) ourselves or even for us to upgrade our scripts to ES6 to be able to support this. Though the latter would be out of scope for this PR.

I agree with that :+1: and will be happy to create it myself when the discussion here about the implementation is solved.

**Comment by ericcornelissen** [comment]
For some in-line responses, see below. As you brought up performance though, I do want to bring up this possibly very naïve point: for a system that does support UTF8, it doesn't seem straightforwardly true to me that decoding an encoded character is necessarily faster than "decoding" a UTF8 character. Especially when the website states it's using UTF8 encoding, because presumably every letter of of the encoded character is now parsed as a UTF8 character.

---

> I think so, but are we targeting websites exclusively?

I don't think that's something that has ever been "formally" decided, though most of our processes suggest it is at least our primary target (npm, packagist, webfont, website).


> More formally, are we targeting UTF-8 encoded based software?

I guess this is the PR we're going to decide on that :upside_down_face: 


> What happens with some people stucked in some other encodings, or in other contexts rather than websites?

That is why I suggested we add a note about it to the docs. That way people _can_ be aware of it and preprocess the SVGs to ASCII themselves.

There's definitely something to say in favor of "just use ASCII to support everyone", but if that were generally true there wouldn't be any other encodings.


> > I don't know if that changes much about the changes in this PR though...
> Just a number in the 127 >=< loops, there is no problem at all.

Not sure if I follow - by my comment I meant "I don't know if changing the character encoding in this Pull Request to UTF8 changes this Pull Request much".


> if you are assuming UTF-8 you can even remove the function `titleToHtmlFriendly`,

As you also said later in your comment, we always need to escape certain characters (at the very least `&`, `>`, and `<`), so we will always need that function in some capacity.


> There is no source because it depends how many encodings do you want "to support". There is no such thing of "what characters should be escaped or should not",

I understand stand that, what I meant to say is that I couldn't really find any definitive source for UTF8 or ASCII in the context of the (modern) web. Or in other words, what is the right thing for us to do under the assumption that we're targeting the web.


> Imagine if I'm a software developer that only wants to support ASCII (one byte UTF-8). This is no a crazy assumption because if you want the fastest SVG path parser in the world, you must forget about other encodings (in fact, you only need a subset of ASCII), assuming that all passed characters are valid ASCII.

I did not intend to imply I think it's "crazy" to go with ASCII, but, again, I do think it warrants a discussion. Not to mention that so far this project has implicitly used UTF8 encoding as accented characters were never encoded.


> That being said, I am more inclined to the preponderance of universal compatibility over "readability of code" when the code is a basic SVG with two readable tags inside and all created from the same basic template.

I think that's a decent argument in favor of just going with ASCII only. Though it should be noted that it's yet another thing that slightly complicates contributing to this project.


> I agree with that :+1: and will be happy to create it myself when the discussion here about the implementation is solved.

Sounds good :+1:

**Comment by mondeja** [comment]
> I think that's a decent argument in favor of just going with ASCII only. Though it should be noted that it's yet another thing that slightly complicates contributing to this project.

If I remember fine, there are 23 icons now with non ASCII characters, which represents 0.0114% of all (I'm taking here 2009 icons). Given that there are a very few icons with non ASCII characters and the the fact that the linter is pretty smart triggering messages indicating the exact replacements that contributors should do I suppose that this would not be problem at all.

Maybe other @simple-icons/maintainers want to share their opinions regardling this topic. If there aren't in some weeks and @ericcornelissen agree, I will continue the work here implementing safer decoding and encoding functions in a separate `all-named-html-entities`-like package which will be moved to simple-icons organization.

File: icons/atandt.svg
Comment (mondeja) [reply icons/atandt.svg:1]: Sorry, I see now. Seems a legit source :+1: So we'll make exceptions in decimal number representation just for `"&<>`.

**Comment by sachinraja** [comment]
We could just use the previous version, `all-named-html-entities@1.6.0`.

**Comment by ericcornelissen** [comment]
> If I remember fine, there are 23 icons now with non ASCII characters, which represents 0.0114% of all (I'm taking here 2009 icons). Given that there are a very few icons with non ASCII characters and the the fact that the linter is pretty smart triggering messages indicating the exact replacements that contributors should do I suppose that this would not be problem at all.

I'm not saying I'm definitely against using ASCII for that reason, just pointing out it is a thing to consider. To clarify, I'm fine with going with ASCII, I'm just trying to make sure we thought about our options before just deciding on an approach without much consider. (Which does not mean I'm saying you didn't consider anything, just that I'm not seeing those considerations in the Pull Request description or commit messages.)

**Comment by service-paradis** [comment]
> Maybe other @simple-icons/maintainers want to share their opinions regardling this topic. If there aren't in some weeks and @ericcornelissen agree, I will continue the work here implementing safer decoding and encoding functions in a separate all-named-html-entities-like package which will be moved to simple-icons organization.

I have not a strong opinion on this one, but I'm not sure if we're really solving a problem here.

> If I remember fine, there are 23 icons now with non ASCII characters, which represents 0.0114% of all (I'm taking here 2009 icons). Given that there are a very few icons with non ASCII characters and the the fact that the linter is pretty smart triggering messages indicating the exact replacements that contributors should do I suppose that this would not be problem at all.

It can also be taken from the other side. I dont know the percentage of people that are using ASCII instead of UTF8, but it should be fairly small. We're doing all this for a small percentage of user for ~0.0114% of our icons. Seems like a huge edge cases.

In cases like this, I'm often more on the side of "If it is not broken, dont fix it" 😅
If there was issues open in this regard, I would be much more in favor.

I'm wondering if it should this be considered a breaking change also? It might be edge cases, but what if someone is parsing the SVG to fetch the title to do something with it. It might not expect encoded characters.

Reading myself, it looks like I'm against it 😅 In fact, I'm not, but I'm not sure if the benefits is real for now.

**Comment by mondeja** [comment]
> I'm wondering if it should this be considered a breaking change also? It might be edge cases, but what if someone is parsing the SVG to fetch the title to do something with it. It might not expect encoded characters.

If someone is parsing HTML, XML or SVG without taking into account encoding... it is just doing a bad work. I think that this is not a breaking change because the outputted icons will be the same.

> It can also be taken from the other side. I dont know the percentage of people that are using ASCII instead of UTF8, but it should be fairly small. We're doing all this for a small percentage of user for ~0.0114% of our icons. Seems like a huge edge cases.

Really interesting because I think exactly the opposite, that this work (a consistent and real encoding implementation) is saving us a lot of future work debugging possible bugs, implementing other things, unifying data, providing support for everyone...

> In cases like this, I'm often more on the side of "If it is not broken, dont fix it"

I agree, the problem comes from that it is broken. This is the reason for which we didn't added tests for svg content. It is just impossible with the current implementation.

**Comment by ericcornelissen** [comment]
> > I'm wondering if it should this be considered a breaking change also? It might be edge cases, but what if someone is parsing the SVG to fetch the title to do something with it. It might not expect encoded characters.
> 
> If someone is parsing HTML, XML or SVG without taking into account encoding... it is just doing a bad work. I think that this is not a breaking change because the outputted icons will be the same.

I agree that we do not have to consider this a breaking change.


> > It can also be taken from the other side. I dont know the percentage of people that are using ASCII instead of UTF8, but it should be fairly small. We're doing all this for a small percentage of user for ~0.0114% of our icons. Seems like a huge edge cases.
> 
> Really interesting because I think exactly the opposite, that this work (a consistent and real encoding implementation) is saving us a lot of future work debugging possible bugs, implementing other things, unifying data, providing support for everyone...

I think both sides are correct here. Now that we're more aware of the fact that we currently largely target UTF8, issues related to encoding should be relatively easy to "debug". If they come up, just ask the reporter what character encoding they're expecting.

That said, I don't think we're doing a lot of work here, we're mostly trying to understand what's the best course of action.


> > In cases like this, I'm often more on the side of "If it is not broken, dont fix it"
> 
> I agree, the problem comes from that it is broken. This is the reason for which we didn't added tests for svg content. It is just impossible with the current implementation.

I think @service-paradis' point was that it's not broken in the sense that no-one ever reported an issue related to this. I don't agree that the tests are "impossible" to do with the current implementation, it's just not sustainable. We would need to keep updating `titleToHtmlFriendly` as new characters are introduced. Plus, the fact that it's pretty easy to miss it when a new character is introduced that should've been encoded is not ideal.

File: icons/cycling74.svg
Comment (ericcornelissen) [reply icons/cycling74.svg:1]: @mondeja, did you see this before? Are you sure `'` and `"` (_normal_ single and double quote) have to be escaped?

File: icons/atandt.svg
Comment (ericcornelissen) [reply icons/atandt.svg:1]: @mondeja, what's the decision for `&<>` (see [this comment](https://github.com/simple-icons/simple-icons/pull/6522#discussion_r716076020) for `"`)? Just asking to be sure everyone's on the same page. I still think using e.g. `&amp;` is "nicer", but numeric is (as you have said elsewhere) easier to implement - so I'd be fine either way.

**Comment by ericcornelissen** [comment]
@mondeja, shall we move ahead with the ASCII-only approach?

File: icons/cycling74.svg
Comment (mondeja) [reply icons/cycling74.svg:1]: I'm not sure what it means that *has to be escaped*. I think that it depends on the usage case. A web application allowing to edit HTML and SVG in an ambiguous context *should* escape it to avoid a security problem, but other usages don't need this.

**Comment by mondeja** [comment]
> @mondeja, shall we move ahead with the ASCII-only approach?

- All non ASCII characters are escaped using their decimal representation `&#[0-9]+;`.
- For 4 ASCII characters, XML named entity references are used: `&amp;` (`&`), `&quot;` (`"`), `&lt;` (`<`) and `&gt;` (`>`).
- For the ASCII character `'`, its decimal representation is used `&#39;`.
- The `named-html-entities-json` package is [here](https://github.com/mondeja/named-html-entities-json). I'll be happy to transfer the repository to simple-icons organization.

File: icons/cycling74.svg
Comment (ericcornelissen) [reply icons/cycling74.svg:1]: By "has to be escaped" I meant in the context of making it HTML safe. But if there's another reason for escaping it then we can still do it. What is the security problem this avoids?

**Review by ericcornelissen — APPROVED** [review]
The lint rule looks good to the extend I can follow it (two minor formatting issues I noticed: L194 extra newline, and L230, extra space before `)`). Let's resolve [this discussion](https://github.com/simple-icons/simple-icons/pull/6522/files#r716970344) before merging

> I'll be happy to transfer the repository to simple-icons organization.

That doesn't seem necessary unless you don't want to maintain it separately from this project. Of course, I'd say you're welcome to if you think it's nicer.

File: icons/cycling74.svg
Comment (mondeja) [reply icons/cycling74.svg:1]: See [Does escaping quotes in a HTML attribute context prevent XSS?](https://security.stackexchange.com/questions/247872/does-escaping-quotes-in-a-html-attribute-context-prevent-xss).

Comparing the behaviour encoding them and not, this will render an empty body in the navigator because is invalid HTML:

```html
<body>
  <svg attr="><title>Hello&quot;</title></svg>
</body>
```

But the next unencoded snippet will render `<svg attr="><title>Hello" <="" title=""></svg>` in the body:

```html
<body>
  <svg attr="><title>Hello"</title></svg>
</body>
```

It seems to me that also prevents malformatted output.

Comment (ericcornelissen) [reply icons/cycling74.svg:1]: I disagree XSS is relevant here. If you're using the simple-icons SVG in the way you illustrated I would say the user of our package is responsible for properly escaping inputs. Not to mention that the full SVG has many double quotes, all of which would cause issue if you're embedding it directly as an attribute.

Comment (mondeja) [reply icons/cycling74.svg:1]: > Not to mention that the full SVG has many double quotes, all of which would cause issue if you're embedding it directly as an attribute.

My point of view is that there are more chances that some parsers will detect quotes in the title as part of the value of an attribute instead of a character in the title. I'm fine also without encoding the character but I feel that it is not intrusive and could prevent potential problems.

Comment (ericcornelissen) [reply icons/cycling74.svg:1]: Can you elaborate @mondeja? I really don't understand what the scenario is you're thinking about. Why would the title be seen as an attribute? it's HTML/XML content since it's in between the `<title>` tags so it can't be an attribute.

If you're thinking about people embedding e.g. `require("icons/icon.js").title` directly into an attribute (i.e., as `<div attr="${require("icons/icon.js").title}"></div>`), then yes there could be a problem. However, changing the `<title>` value in the SVG doesn't help here. Moreover, depending on where it's embedded it should be escaped differently. E.g., if you're embedding it in this way in a JavaScript string, then the title value should be `Hello\\"` (or something similar).

I'm also not necessarily against encoding it, it's not a huge deal as the raw title in the SVG isn't really meant to be read by a human. (I do still think this discussion is useful as it provides context to anyone trying to understand why these characters are encoded in the future.)

Comment (mondeja) [reply icons/cycling74.svg:1]: Yes, I was thinking in users inserting parts of the SVGs in their HTML contents, but it would be a rare usage case. We could choose not to encode these characters and implement encoding later for them if any related issue arises, do you agree?

Comment (ericcornelissen) [reply icons/cycling74.svg:1]: That sounds good to me @mondeja 👍

Comment (mondeja) [reply icons/cycling74.svg]: Done in https://github.com/simple-icons/simple-icons/pull/6522/commits/5561224c9c03c299cd8b39514d3265e4eefd138d

**Review by ericcornelissen — APPROVED** [review]

**Comment by ericcornelissen** [comment]
Thank you for your work on this one @mondeja, once again an important improvement to the project!

**Pull request merged**
